runtime/near-wallet-contract/build.rs

```diff
@@ -0,0 +1,65 @@
+/// This file is run as a part of `cargo build` process and it builds the `Wallet Contract`.
+/// The generated WASM file is put to the `./res` directory.
+use anyhow::{anyhow, Context, Ok, Result};
+
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+fn main() -> Result<()> {
+    if cfg!(not(feature = "nightly")) {
+        return Ok(());
+    }
+    build_contract("./wallet-contract", &[], "wallet_contract")
+}
+
+fn build_contract(dir: &str, args: &[&str], output: &str) -> Result<()> {
+    let target_dir: PathBuf =
+        std::env::var("OUT_DIR").context("Failed to read OUT_DIR environment variable")?.into();
+
+    // We place the build artifacts in `target_dir` (workspace's build directory).
+    let mut cmd = cargo_build_cmd(&target_dir);
+    cmd.args(args);
+    cmd.current_dir(dir);
+    run_checking_status(cmd)?;
+
+    let build_artifact_path =
+        format!("wasm32-unknown-unknown/release/{}.wasm", dir.replace('-', "_"));
+    let src = target_dir.join(build_artifact_path);
+    let wasm_target_path = format!("./res/{}.wasm", output);
+
+    std::fs::copy(&src, &wasm_target_path)
+        .with_context(|| format!("Failed to copy `{}` to `{}`", src.display(), wasm_target_path))?;
+
+    println!("cargo:rerun-if-changed={}", dir);
+    Ok(())
+}
+
+/// Creates `cargo build` command to compile the WASM file.
+/// Note that we are in `build.rs` file, so this will be called as a part
+/// of the global `cargo build` process that already has some flags set.
+/// `env_remove` invocations will remove these flags from the nested `cargo build`
+/// process, to avoid unexpected behaviors due to the workspace configurations.
+fn cargo_build_cmd(target_dir: &Path) -> Command {
```

Comment:
The way this contract is built will need to be changed for it to have a reproducible hash. By default Rust compilation includes things like the absolute paths for the source code compiled, which obviously differ between machines. Near publishes a [Docker container](https://hub.docker.com/r/nearprotocol/contract-builder) with a standardized build environment to achieve reproducible builds.

We can handle this in a follow-up PR though I suppose.

Reply:
There’s the `--remap-path-prefix` option for rustc that should enable reproducible builds.

Reply:
Thank you @nagisa! I was not aware of that.